### PR TITLE
make pull --run-cache work with local remote having relative path

### LINKED ISSUE #4637: `dvc pull --run-cache` does not work if remote path is a relative path
## Bug Report

`dvc push --run-cache` or plain `dvc push/pull` works though.
### Please provide information about your setup

**Additional Information (if any):**
```console
❯ dvc pull --run-cache -v
2020-09-29 16:50:06,143 DEBUG: Check for update is enabled.
2020-09-29 16:50:06,148 DEBUG: fetched: [(3,)]
2020-09-29 16:50:06,149 DEBUG: fetched: [(6,)]
2020-09-29 16:50:06,230 ERROR: unexpected error - '../storage/runs/f4/f4d96ea9e3a842c243baf6a9ba72c2e95cf67d129159eafb813efa6859aa0a59/09eac8a46ebb889acde085a8ec7d617deb9179ae54908323ca451ce7a3612f8c' does not start with '/home/saugat/dvc_try/stage_cache/folder1/.dvc/../../storage'
------------------------------------------------------------
Traceback (most recent call last):
  File "/home/saugat/repos/iterative/dvc/dvc/main.py", line 76, in main
    ret = cmd.run()
  File "/home/saugat/repos/iterative/dvc/dvc/command/data_sync.py", line 26, in run
    stats = self.repo.pull(
  File "/home/saugat/repos/iterative/dvc/dvc/repo/__init__.py", line 51, in wrapper
    return f(repo, *args, **kwargs)
  File "/home/saugat/repos/iterative/dvc/dvc/repo/pull.py", line 25, in pull
    processed_files_count = self.fetch(
  File "/home/saugat/repos/iterative/dvc/dvc/repo/__init__.py", line 51, in wrapper
    return f(repo, *args, **kwargs)
  File "/home/saugat/repos/iterative/dvc/dvc/repo/fetch.py", line 39, in fetch
    used_run_cache = self.stage_cache.pull(remote) if run_cache else []
  File "/home/saugat/repos/iterative/dvc/dvc/stage/cache.py", line 217, in pull
    return self._transfer(
  File "/home/saugat/repos/iterative/dvc/dvc/stage/cache.py", line 196, in _transfer
    rel = src.relative_to(from_remote.path_info)
  File "/usr/lib/python3.8/pathlib.py", line 904, in relative_to
    raise ValueError("{!r} does not start with {!r}"
ValueError: '../storage/runs/f4/f4d96ea9e3a842c243baf6a9ba72c2e95cf67d129159eafb813efa6859aa0a59/09eac8a46ebb889acde085a8ec7d617deb9179ae54908323ca451ce7a3612f8c' does not start with '/home/saugat/dvc_try/stage_cache/folder1/.dvc/../../storage'
------------------------------------------------------------

Having any troubles? Hit us up at https://dvc.org/support, we are always happy to help!
DVC version: 1.7.9+bb5b2a.mod
---------------------------------
Platform: Python 3.8.5 on Linux-5.8.11-arch1-1-x86_64-with-glibc2.2.5
Supports: All remotes
Cache types: hardlink, symlink
Cache directory: ext4 on /dev/sda9
Workspace directory: ext4 on /dev/sda9
Repo: dvc, git
2020-09-29 16:50:06,898 DEBUG: Analytics is enabled.
2020-09-29 16:50:07,000 DEBUG: Trying to spawn '['daemon', '-q', 'analytics', '/tmp/tmp8vs2s28d']'
2020-09-29 16:50:07,003 DEBUG: Spawned '['daemon', '-q', 'analytics', '/tmp/tmp8vs2s28d']'

```

#### Example Config
```ini
[core]
    remote = local
['remote "local"']
    url = ../../storage

```

### PR DESCRIPTION
**Pull request by musa-atlihan**
* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

Fixes #4637.

### CODE REVIEW

File: dvc/stage/cache.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 import os
 import tempfile
 from contextlib import contextmanager
+from pathlib import Path
```
Comment (pmrowla) [inline dvc/stage/cache.py:5]:
```suggestion
```

(see related comment below)

File: dvc/stage/cache.py [hunk 2]
```diff
@@ -197,7 +198,7 @@ def _transfer(func, from_remote, to_remote):
             return []
 
         for src in from_remote.walk_files(runs):
-            rel = src.relative_to(from_remote.path_info)
+            rel = Path(os.path.relpath(src, from_remote.path_info))
```
Comment (pmrowla) [inline dvc/stage/cache.py:201]:
```suggestion
            rel = PathInfo(os.path.relpath(src, from_remote.path_info))
```

DVC `PathInfo` inherits from python's `PurePath` and should be preferred here

**Review by pmrowla — CHANGES_REQUESTED** [review]
Looks like this is related to https://bugs.python.org/issue40358 `Path.relative_to` behaves differently than `os.path.relpath` in this situation

The changes in this PR do resolve the original bug, but I think it would be better for us to catch the `ValueError` in `PathInfo.relative_to` when we call the parent (python) `relative_to` function, and fallback to `os.path.relpath` there
https://github.com/iterative/dvc/blob/15b1e7072bf4c133a640507f59011d32501a968f/dvc/path_info.py#L227

that way we won't end up running into this same bug elsewhere in DVC

**Comment by musa-atlihan** [comment]
thank you for the review @pmrowla ! I will update and push now.

**Comment by pmrowla** [comment]
@musa-atlihan have you looked into trying my other suggestion regarding fixing `PathInfo.relative_to` in general (so that we handle this issue everywhere in DVC and not just this specific run cache bug)?

**Comment by musa-atlihan** [comment]
@pmrowla , I pushed for the `PathInfo.relative_to` fix, could you check?

File: dvc/stage/cache.py [hunk 3]
```diff
@@ -197,7 +197,7 @@ def _transfer(func, from_remote, to_remote):
             return []
 
         for src in from_remote.walk_files(runs):
-            rel = src.relative_to(from_remote.path_info)
+            rel = PathInfo(os.path.relpath(src, from_remote.path_info))
```
Comment (pmrowla) [inline dvc/stage/cache.py:200]: I think this change is no longer needed after the fix in `path_info.py`, @musa-atlihan can you try reverting this and then confirm that `pull --run-cache` works as expected?

**Comment by pmrowla** [comment]
Thank you for the PR @musa-atlihan! 🙏

**Pull request merged**